### PR TITLE
Fix Robolectric tests failed to run

### PR DESCRIPTION
**Pull request by TranceLove**
- Added monkey-patched `ShadowMultiDex` for supporting `MultiDex` from Android Jetpack
- Limit `maxSdk` to 27 (as of Robolectric 3.8)
- Update code to match updated APIs

Some tests are skipped.
- `MainActivityTest` skipped since it does nothing
- `TextEditorActivityTest` restricted to SDK 24-27 due to [possibly bug within Robolectric 3.8](https://github.com/robolectric/robolectric/issues/3947), causing tests to fail in SDK 21-23
- `SevenZipHelperTaskTest2` skipped due to bug at 7zip extraction library we forked from commons-compress. 7zip file in question was not extractable in current master anyway...

### CODE REVIEW

**Comment by VishalNehra** [comment]
Unable to compile, ShadowMultiDex not found. Am I missing something.

**Comment by TranceLove** [comment]
> Unable to compile, ShadowMultiDex not found. Am I missing something.

Sorry...

**Comment by EmmanuelMess** [comment]
Getting:

> java.lang.NullPointerException
> 	at org.robolectric.internal.DefaultManifestFactory.getFsFileFromPath(DefaultManifestFactory.java:65)
> 	at org.robolectric.internal.DefaultManifestFactory.identify(DefaultManifestFactory.java:24)
> 	at org.robolectric.RobolectricTestRunner.getAppManifest(RobolectricTestRunner.java:431)
> 	at org.robolectric.RobolectricTestRunner.getChildren(RobolectricTestRunner.java:255)
> 	at org.junit.runners.ParentRunner.getFilteredChildren(ParentRunner.java:426)
> 	at org.junit.runners.ParentRunner.filter(ParentRunner.java:379)
> 	at org.junit.runner.manipulation.Filter.apply(Filter.java:97)
> 	at org.junit.internal.requests.FilterRequest.getRunner(FilterRequest.java:37)
> 	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:49)
> 	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
> 	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
> 	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

**Comment by EmmanuelMess** [comment]
Robolectric crashes trying to find tests to run:

```java
java.lang.NullPointerException
	at org.robolectric.internal.DefaultManifestFactory.getFsFileFromPath(DefaultManifestFactory.java:65)
	at org.robolectric.internal.DefaultManifestFactory.identify(DefaultManifestFactory.java:24)
	at org.robolectric.RobolectricTestRunner.getAppManifest(RobolectricTestRunner.java:431)
	at org.robolectric.RobolectricTestRunner.getChildren(RobolectricTestRunner.java:255)
	at org.junit.runners.ParentRunner.getFilteredChildren(ParentRunner.java:426)
	at org.junit.runners.ParentRunner.getDescription(ParentRunner.java:351)
	at com.intellij.junit4.JUnit4IdeaTestRunner.getDescription(JUnit4IdeaTestRunner.java:78)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:50)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by TranceLove** [comment]
> Robolectric crashes trying to find tests to run:
> 
> ```java
> java.lang.NullPointerException
> 	at org.robolectric.internal.DefaultManifestFactory.getFsFileFromPath(DefaultManifestFactory.java:65)
> 	at org.robolectric.internal.DefaultManifestFactory.identify(DefaultManifestFactory.java:24)
> 	at org.robolectric.RobolectricTestRunner.getAppManifest(RobolectricTestRunner.java:431)
> 	at org.robolectric.RobolectricTestRunner.getChildren(RobolectricTestRunner.java:255)
> 	at org.junit.runners.ParentRunner.getFilteredChildren(ParentRunner.java:426)
> 	at org.junit.runners.ParentRunner.getDescription(ParentRunner.java:351)
> 	at com.intellij.junit4.JUnit4IdeaTestRunner.getDescription(JUnit4IdeaTestRunner.java:78)
> 	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:50)
> 	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
> 	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
> 	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
> ```

Sorry again, forgotten to push changed gradle.properties too. See 2e72bb8.

Workaround was from [here](https://github.com/robolectric/robolectric/issues/4690#issuecomment-547574085).

**Comment by EmmanuelMess** [comment]
`UsbOtgTest` and `SigletonUsbOtgTest` throw NPE. And there are quite a few `java.lang.UnsupportedOperationException: Robolectric does not support API level 28.`, if you ignore those, I will approve, so that we have actually running tests.

**Comment by VishalNehra** [comment]
I'm getting various test fails on IconsTest
```com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[17] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[18] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[19] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[21] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[22] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[23] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[24] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[25] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException

com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[26] FAILED
    org.apache.tools.ant.BuildException
        Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
```

**Comment by TranceLove** [comment]
> I'm getting various test fails on IconsTest
> 
> ```
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[18] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[19] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[21] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[22] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[23] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[24] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[25] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> 
> com.amaze.filemanager.ui.icons.IconsTest > testReturnArchiveTypes[26] FAILED
>     org.apache.tools.ant.BuildException
>         Caused by: org.apache.maven.artifact.resolver.MultipleArtifactsNotFoundException
> ```

It seems to be a build failure... did you `clean` before running the tests?

**Comment by EmmanuelMess** [comment]
@TranceLove I cannot get the tests to run, it says classes are missing, the test classes, even after clean and rebuild.

**Comment by TranceLove** [comment]
> @TranceLove I cannot get the tests to run, it says classes are missing, the test classes, even after clean and rebuild.

even with 2aa5c97?

I did tried using command line `gradlew` to run the tests, and I'm getting `OutOfMemoryError` when running `com.amaze.filemanager.asynchronous.services.ExtractServiceTest`. Will see what happened.

**Comment by VishalNehra** [comment]
I am getting same errors on IconsTest even after clean build. I used gradle wrapper only.

**Comment by EmmanuelMess** [comment]
Any news?

**Comment by TranceLove** [comment]
> Any news?

Anything I missed?

**Comment by EmmanuelMess** [comment]
> > Any news?
> 
> Anything I missed?

ShadowMultidex import is broken

**Comment by TranceLove** [comment]
Build artifact

**Comment by VishalNehra** [comment]
Changes seems fine, please check few points before we can merge this
0. Stuck on com.amaze.filemanager.filesystem.ssh.ListFilesOnSshdTest (308 tests completed) from 45 mins. Can we maybe move integration tests like this outside the scope of this repo or ignore for the time being.
1. Rebase and point to release/3.5 instead of master
2. Codacy showing minor import issues (might be fixed automatically once you rebase with 3.5 and perform spotlessApply). Also I've ignored `commons_compress_7z` from codacy, not sure why it's showing there, maybe after rebase will be fixed.
3. `PemToKeyPairTaskTest2` seem to have missing imports.
4. Add jacoco to module's `build.gradle` so we can configure codacy to show test case coverage.
```apply plugin: 'jacoco'```
and 
```
buildTypes {
        debug {
            ...
            testCoverageEnabled true
            ...
        }
    }
```

**Comment by TranceLove** [comment]
Espresso tests fixed missing imports... and also enabled test coverage in 595ed82.

The errors you got during unit tests didn't happen on my environments, though the long test time was indeed a problem. Will try again using Docker image for a fairer comparison.

**Comment by VishalNehra** [comment]
In that case, let's push this to release/3.5 then. 
I forgot to request changes to add test step in pipeline. Can you please add them in .github workflows as well? Can exclude this from artifact workflows IMO.

**Comment by TranceLove** [comment]
Test build apk

**Comment by TranceLove** [comment]
Build test apk after unit test

File: .github/workflows/android-debug-artifact-ondemand.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ jobs:
       - name: Build with Gradle
         run: |
           echo Run on-demand by ${GITHUB_ACTOR}
-          bash ./gradlew assembleDebug --stacktrace
+          bash ./gradlew testPlayDebugUnitTest assembleDebug --stacktrace
```
Comment (VishalNehra) [inline .github/workflows/android-debug-artifact-ondemand.yml:41]: Please remove `testPlayDebugUnitTest` from here as apk generation doesn't necessarily require test cases to pass.

File: .github/workflows/android-main.yml [hunk 2]
```diff
@@ -21,7 +21,7 @@ jobs:
           java-version: 1.8
       - name: Build with Gradle
         run: |
-          ./gradlew assembledebug
+          ./gradlew testPlayDebugUnitTest assembledebug
```
Comment (VishalNehra) [inline .github/workflows/android-main.yml:24]: Please add flag `--stacktrace` as well.

File: .github/workflows/android-feature.yml [hunk 3]
```diff
@@ -18,7 +18,7 @@ jobs:
           java-version: 1.8
       - name: Build with Gradle
         run: |
-          ./gradlew assembledebug
+          ./gradlew testPlayDebugUnitTest assembledebug
```
Comment (VishalNehra) [inline .github/workflows/android-feature.yml:21]: Please add flag `--stacktrace` as well.

File: README.md [hunk 4]
```diff
@@ -65,6 +65,7 @@ Authors
 
     Copyright (C) 2014-2018 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>
```
Comment (VishalNehra) [inline README.md:66]: Can you please separate out my name as well, and change it to 2014-2020 :P

**Pull request merged**